ksql-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java

```diff
@@ -200,44 +211,294 @@ SchemaKTable tableForJoin(
         functionRegistry,
         joinTableProps, schemaRegistryClient);
     if (!(schemaKStream instanceof SchemaKTable)) {
-      throw new KsqlException("Unsupported Join. Only stream-table joins are supported, but was "
-          + getLeft() + "-" + getRight());
+      throw new KsqlException("Expected " + getSourceName(right) + " to be a table, but found a "
+                              + "stream instead.");
     }
 
     return (SchemaKTable) schemaKStream;
   }
 
 
-  private KsqlTopicSerDe getResultTopicSerde(final PlanNode node) {
-    if (node instanceof StructuredDataSourceNode) {
-      StructuredDataSourceNode structuredDataSourceNode = (StructuredDataSourceNode) node;
-      return structuredDataSourceNode.getStructuredDataSource().getKsqlTopic().getKsqlTopicSerDe();
-    } else if (node instanceof JoinNode) {
-      JoinNode joinNode = (JoinNode) node;
+  private void ensureMatchingPartitionCounts(final KafkaTopicClient kafkaTopicClient) {
+    final int leftPartitions = left.getPartitions(kafkaTopicClient);
+    final int rightPartitions = right.getPartitions(kafkaTopicClient);
 
-      return getResultTopicSerde(joinNode.getLeft());
-    } else {
-      return getResultTopicSerde(node.getSources().get(0));
+    if (leftPartitions != rightPartitions) {
+      throw new KsqlException("Can't join " + getSourceName(left) + " with "
+                              + getSourceName(right) + " since the number of partitions don't "
+                              + "match. " + getSourceName(left) + " partitions = "
+                              + leftPartitions + "; " + getSourceName(right) + " partitions = "
+                              + rightPartitions + ". Please repartition either one so that the "
+                              + "number of partitions match.");
     }
   }
 
-  @Override
-  protected int getPartitions(KafkaTopicClient kafkaTopicClient) {
-    return right.getPartitions(kafkaTopicClient);
+  private String getSourceName(PlanNode node) {
+    if (!(node instanceof StructuredDataSourceNode)) {
+      throw new KsqlException("The source for a join must be a Stream or a Table.");
```

Comment:
Is this needed? Seems that this method only is called for creating an exception message and if you have this exception, it would supersede the other exception message!

Reply:
The only reason for this method is support printing a nicer error message which name each source in case the number of partitions don't match. I changed the exception to be an instance of `RunTimeException`, since it is just a sanity check and should never be raised unless there is bug.